In [1]:
from sklearn.cluster import KMeans
from Coreset import *
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import normalized_mutual_info_score as NMI
import numpy as np
from tqdm import tqdm
import time

In [2]:
data = np.genfromtxt("data/bio_train.csv", delimiter=",")[:,3:]
data = MinMaxScaler().fit_transform(data)
labels_true = np.genfromtxt("data/bio_train.csv", delimiter=",")[:,0]

In [5]:
ms = [145750, 112991, 86814, 68783, 56954]

for m in tqdm(ms):

    nmis = []
    accuracies = []
    iters = []
    start = time.time()

    for _ in tqdm(range(10)):
        cs = Coreset(data)
        cs.get_q()
        coreset, rnd_labels = cs.get_coreset(m)

        kmeans = KMeans(n_clusters=153)
        kmeans.fit(data[rnd_labels], sample_weight=cs.weights)
        predict = kmeans.predict(data)
        nmis.append(NMI(labels_true, predict))
        iters.append(kmeans.n_iter_)
        accuracies.append(np.mean(labels_true == predict))

    print(f'\n----Results for Coreset Test using size {m}----\n')
    print(f'Average Accuracy over 10 runs: {np.mean(nmis):.3f}')
    print(f'Variance over 10 runs: {np.var(accuracies):.3f}')
    print(f'Average number of iterations: {np.mean(iters)}')
    print(f'Time taken: {time.time() - start:.3f}s')


  0%|          | 0/10 [00:00<?, ?it/s]
----Results for Coreset Test using size 145750----

Average Accuracy over 10 runs: 0.202
Variance over 10 runs: 0.000
Average number of iterations: 213.0
Time taken: 2260.085s

  0%|          | 0/10 [00:00<?, ?it/s]
----Results for Coreset Test using size 112991----

Average Accuracy over 10 runs: 0.202
Variance over 10 runs: 0.000
Average number of iterations: 164.6
Time taken: 1653.606s

  0%|          | 0/10 [00:00<?, ?it/s]
----Results for Coreset Test using size 86814----

Average Accuracy over 10 runs: 0.202
Variance over 10 runs: 0.000
Average number of iterations: 165.6
Time taken: 1175.382s

  0%|          | 0/10 [00:00<?, ?it/s]
----Results for Coreset Test using size 68783----

Average Accuracy over 10 runs: 0.201
Variance over 10 runs: 0.000
Average number of iterations: 152.7
Time taken: 891.256s

100%|██████████| 5/5 [1:50:52<00:00, 1330.59s/it]
----Results for Coreset Test using size 56954----

Average Accuracy over 10 runs: 0.201
V

## ----Results for Coreset Test using size 145750----

Average Accuracy over 10 runs: 0.202 |
Variance over 10 runs: 0.000 |
Average number of iterations: 213.0 |
Time taken: 2260.085s

## ----Results for Coreset Test using size 112991----

Average Accuracy over 10 runs: 0.202 |
Variance over 10 runs: 0.000 |
Average number of iterations: 164.6 |
Time taken: 1653.606s

## ----Results for Coreset Test using size 86814----

Average Accuracy over 10 runs: 0.202 |
Variance over 10 runs: 0.000 |
Average number of iterations: 165.6 |
Time taken: 1175.382s

## ----Results for Coreset Test using size 68783----

Average Accuracy over 10 runs: 0.201 |
Variance over 10 runs: 0.000 |
Average number of iterations: 152.7 |
Time taken: 891.256s

## ----Results for Coreset Test using size 56954----

Average Accuracy over 10 runs: 0.201 |
Variance over 10 runs: 0.000 |
Average number of iterations: 142.6 |
Time taken: 672.633s